tests/41end-to-end-keys/06-device-lists.pl

```diff
@@ -9,13 +9,34 @@ sub is_user_in_changed_list
           any { $_ eq $user->user_id } @{ $body->{device_lists}{changed} };
 }
 
-sub is_user_in_left_list
+
+# returns a Future which resolves to the body of the sync result which contains
+# the change notification
+sub sync_until_user_in_device_list
 {
-   my ( $user, $body ) = @_;
+   my ( $syncing_user, $user_to_wait_for, %params ) = @_;
 
-   return $body->{device_lists} &&
-          $body->{device_lists}{left} &&
-          any { $_ eq $user->user_id } @{ $body->{device_lists}{left} };
+   my $device_list = $params{device_list} // 'changed';
+   my $msg = $params{msg} // 'sync_until_user_in_device_list';
+
+   my $wait_for_id = $user_to_wait_for->user_id;
+
+   $msg = "$msg: waiting for $wait_for_id in $device_list";
+
+   return repeat_until_true {
+      matrix_sync_again( $syncing_user, timeout => 1000 )
+      ->then( sub {
+         my ( $body ) = @_;
+
+         log_if_fail "$msg: body", $body;
+
+         my $res = $body->{device_lists} &&
+            $body->{device_lists}{$device_list} &&
+            any { $_ eq $wait_for_id } @{ $body->{device_lists}{$device_list} };
+
+         Future->done( $res && $body );
```

Comment:
Why the ` && $body`?

Reply:
to make it return the body, as per https://github.com/matrix-org/sytest/pull/387/files#diff-f4f7b7565a33f2c8f4cc0f2c24bb75f6R13

Reply:
oh yes thats how this language works